packages/components/src/progressive-disclosure-panel/index.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	createContext,
+	useContext,
+	useEffect,
+	useMemo,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import ProgressiveDisclosurePanelItem from './item';
+import ProgressiveDisclosurePanelTitle from './title';
+
+const PanelContext = createContext( {} );
+
+export const usePanelContext = () => useContext( PanelContext );
+
+const isMenuItem = ( item ) => item.type === ProgressiveDisclosurePanelItem;
+
+const ProgressiveDisclosurePanel = ( props ) => {
+	const { children, className, label: menuLabel, resetAll, title } = props;
+	const [ menuItems, setMenuItems ] = useState( {} );
```

Comment:
Just a thought the approach ended up not simpler so feel free to skip:
What if instead of the parent knowing which elements/menu items are toggled on or off, each one keeps track of itself inside ProgressiveDisclosurePanelItem. Basically, we have a boolean state inside ProgressiveDisclosurePanelItem that says if we should force show that item. The checks if an items hasValue etc. would also all be done inside ProgressiveDisclosurePanelItem. Each item would compute if it should render its UI and if the MenuItem should be toggled on or off.
The parent ProgressiveDisclosurePanel would not have any state context etc. it would just render the children and provide a slot for menu items where each ProgressiveDisclosurePanelItem can render the fills (each menu item toggled on or off).

I think this approach would totally remove the need to inspect props and even the need for context.

What complicates things is the reset all button. For this to work, instead of passing just a react element to the slot with the menu, we would need to pass an object that contains the menu element and a function to reset its state. The slot would just render each item and would create a reset all function that essentially calls in a loop all the reset functions.


The reset all complicates the approach I had in mind, so it may not be worth pursuing it (when I started writing, I did not remember the reset all). I'm sharing here anyway in case it inspires something I'm not considering.

Reply:
I ran into a similar wall with the `resetAll` option. A simple loop over individual reset callbacks has a number of problems with it. Mainly, overwriting previous resets and then increasing complexity to try and work around that. 

It was the reason I settled upon a separate `resetAll` callback. Handling the menu state etc at the panel level then seemed to make more sense.